### PR TITLE
[Promotion] Coupon generation refactoring

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Related tickets | mentioned in [comment](https://github.com/Sylius/Sylius/pull/5087#issuecomment-220919677)
| License         | MIT

Based on #5087

### CODE REVIEW

File: src/Sylius/Component/Promotion/Generator/CouponGeneratorInterface.php [hunk 1]
```diff
@@ -24,6 +24,8 @@
     /**
      * @param PromotionInterface $promotion
      * @param InstructionInterface $instruction
+     * 
+     * @return array
```
Comment (pamil) [inline src/Sylius/Component/Promotion/Generator/CouponGeneratorInterface.php]: `@return array Generated coupons as keys`?

File: src/Sylius/Component/Promotion/Generator/CouponGenerator.php [hunk 2]
```diff
@@ -72,53 +67,54 @@ public function __construct(
      */
     public function generate(PromotionInterface $promotion, InstructionInterface $instruction)
     {
-        $this->generatedCoupons = [];
+        $generatedCoupons = [];
 
         $this->assertGenerationIsPossible($instruction);
         for ($i = 0, $amount = $instruction->getAmount(); $i < $amount; ++$i) {
-            $code = $this->generateUniqueCode($instruction->getCodeLength());
+            $code = $this->generateUniqueCode($instruction->getCodeLength(), $coupons);
```
Comment (pamil) [inline src/Sylius/Component/Promotion/Generator/CouponGenerator.php]: There is no `$coupons` variable defined

File: src/Sylius/Component/Promotion/Generator/CouponGenerator.php [hunk 3]
```diff
@@ -72,53 +67,54 @@ public function __construct(
...
-
     /**
-     * @param int $codeLength
+     * @param string $codeLength
+     * @param array $coupons
      *
      * @return string
      */
-    protected function generateUniqueCode($codeLength)
+    protected function generateUniqueCode($codeLength, array $coupons)
```
Comment (pamil) [inline src/Sylius/Component/Promotion/Generator/CouponGenerator.php]: `$coupons` -> `$generatedCoupons` / `$alreadyGeneratedCoupons`?

File: src/Sylius/Component/Promotion/Generator/CouponGenerator.php [hunk 4]
```diff
@@ -72,53 +67,54 @@ public function __construct(
...
             $code = strtoupper(substr($hash, 0, $codeLength));
-        } while ($this->isUsedCode($code));
+        } while ($this->isUsedCode($code, $coupons));
 
         return $code;
     }
 
     /**
-     * @param string $code
+     * @param $code
```
Comment (pamil) [inline src/Sylius/Component/Promotion/Generator/CouponGenerator.php]: `@param string $code`

**Comment by lchrusciel** [comment]
@peteward sorry, but this PR will finally fix a BC, which was introduced in #3339.

File: src/Sylius/Component/Promotion/Generator/CouponGenerator.php [hunk 5]
```diff
@@ -72,53 +67,54 @@ public function __construct(
...
 
-
     /**
      * @param int $codeLength
+     * @param array $reservedCoupons
      *
      * @return string
      */
-    protected function generateUniqueCode($codeLength)
+    protected function generateUniqueCode($codeLength, array $reservedCoupons)
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Promotion/Generator/CouponGenerator.php]: Why not `generatedCoupons`? Using the same array under different name within the same class introduces confusion.

**Comment by pjedrzejewski** [comment]
Thank you Łukasz!

**Pull request merged**